### PR TITLE
Fixes #4317, GLFW mouse buttons mapping.

### LINKED ISSUE #4317: Swapped right and middle mouse buttons in GLFW
GLFW defines middle and right mouse buttons constants as integers 2 and 1
https://github.com/kripken/emscripten/blob/master/system/include/GLFW/glfw3.h#L432

But JS events use different integers for it:
1 : Wheel button or middle button (if present)
2 : Right mouse button
http://www.w3schools.com/jsref/event_button.asp

And emscripten's GLFW implementation uses those integers directly without any conversion:
https://github.com/kripken/emscripten/blob/master/src/library_glfw.js#L444

So porting code which use GLFW_MOUSE_BUTTON_RIGHT or GLFW_MOUSE_BUTTON_MIDDLE result in swapped mouse buttons.

### PR DESCRIPTION
**Pull request by guillaumeblanc**
Fixes #4317. Buttons mapping was actually done in onMouseButtonChanged, but not on onMouseButtonDown and onMouseButtonUp.

### CODE REVIEW

**Comment by kripken** [comment]
Looks like this has a merge conflict that needs to be resolved.

Can the existing glfw tests be added to, to check this? (in `tests/test_browser.py` and/or `tests/test_interactive.py`, test names will have glfw in them)

**Comment by guillaumeblanc** [comment]
There are conflicts because I didn't do the changes on the latest submit. Is this a problem?

I looked for glfw mouse tests before sending the PR, but didn't find any, not even in the 2 files you listed. Maybe it can be inspired by test_sdl_moousewheel.c? I'm not very confortable with how these tests work though.

**Comment by guillaumeblanc** [comment]
I have a test that highlights the isssue, based on test_sdl_moousewheel.c. See : [test_glfw_mouse.c.txt](https://github.com/kripken/emscripten/files/549038/test_glfw_mouse.c.txt).

I have a question to make it more robust though. If you look at the main function, I found that I need a delay between calls to sendEvent. I assume it's because the main loop needs to be called in between sendEvent. Is there a way to run the main loop (or any c function actually) from the EM_ASM section?

Regards,
Guillaume

**Comment by kripken** [comment]
(Yeah, merging incoming as you did is ok to fix the merge conflict - we'll squash before merging this PR, so it will go away anyhow.)

Hmm, I do see mouse button checks in `tests/glfw.c, glfw_events.c, glfw3.c, glfw3_events.c`?

Regarding the robustness, yes, that is a problem - there is no way on the web to run the main event loop except to stop running code. So you need to be async. However, instead of a delay, you can fire the next part of the test when the current part finishes. So you wait async for the mouse event, then handle it and verify that part of the test, then send an event for the next part. Alternatively, you can fire all the events with no delay and accumulate the results, which is ok if they don't depend on each other - looks like that's how the tests mentioned above work.

**Comment by guillaumeblanc** [comment]
You're right, there are some mouse button checks already that I didn't new about:
 - glfw.c/glfw3.c: It only tests GLFW_MOUSE_BUTTON_LEFT which was already working ok. Other printf display a button index which can't really help finding the issue.
- glfw3_events.c: The test works because glfwSetMouseButtonCallback code path was already doing the buttons mapping. This file should probably removed now glfw_events.c is supposed to be used for glfw 2 and 3.
- glfw_events.c: The glfwGetMouseButton test could fit well in this file. Tell me if you prefer to have the unit tests there instead of in the file I added. It would actually fix robustness issue as well.

**Comment by guillaumeblanc** [comment]
Are you fine with the latest changes of this PR ?
Any idea about when it can reach a release ?

Thanks,
Guillaume

**Comment by kripken** [comment]
Looks good, however, `browser.test_glfw_events` fails for me?

**Comment by kripken** [comment]
(Regarding the next release, we hope to merge to master in a week or two.)

**Comment by guillaumeblanc** [comment]
browser.test_glfw_events is fixed now.

There's still an issue with Chrome that fails with the key tests (it think it was already the case before which confused me when I submitted my first implementation). The injectKeyEvent implementation doesn't work with Chrome. I don't know js but I tried a few things without sucess. For example this fails also with chrome (received keyCode is always 0):
`            
                var event = new KeyboardEvent(type, {
                    'view': window,
                    'bubbles': true,
                    'cancelable': true,
                    'keyCode': keyCode
                });
                window.dispatchEvent(event);
`

I looked at other examples like test_sdl_key, but they fail also with Chrome.
Any idea about how to fix that for all browsers?

I also submitted a fix (and test) for an error with glfw2 arrow keys mapping.

File: src/library_glfw.js [hunk 1]
```diff
@@ -208,10 +208,10 @@ var LibraryGLFW = {
         case 0x86:return (256+24); // DOM_VK_F23 -> GLFW_KEY_F23
         case 0x87:return (256+25); // DOM_VK_F24 -> GLFW_KEY_F24
         case 0x88:return (256+26); // 0x88 (not used?) -> GLFW_KEY_F25
-        case 0x27:return (256+27); // DOM_VK_RIGHT -> GLFW_KEY_RIGHT
```
Comment (kripken) [inline src/library_glfw.js]: are these changes a glfw vs glfw3 version issue? or was this always wrong?

**Comment by guillaumeblanc** [comment]
The issue was introduced with bb63709 (Add GLFW 3.0 emulation). It was correct before that.

**Comment by kripken** [comment]
Got it. Thanks for the fixes :)

**Pull request merged**